[PyTorch Distributed] Consolidate NCCL_DESYNC_DEBUG and TORCH_DISTRIBUTED_DEBUG=INFO

kwen2501 opened the pull request:
Summary: Infer desync debug from whether TORCH_DISTRIBUTED_DEBUG >= INFO

Test Plan:
1. When TORCH_DISTRIBUTED_DEBUG=INFO:
1.1 Catch mismatched collectives (e.g. broadcast vs reduce) - passed
1.2 Catch mismatched collective sizes - passed
2. QPS test: no performance regression - passed

Differential Revision: D34232827

rohan-varma commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -540,7 +537,12 @@ ProcessGroupNCCL::ProcessGroupNCCL(
   asyncErrorHandling_ = parseEnvVarFlag(NCCL_ASYNC_ERROR_HANDLING);
-  desyncDebug_ = parseEnvVarFlag(NCCL_DESYNC_DEBUG);
+  // Infer desync debug from whether TORCH_DISTRIBUTED_DEBUG >= INFO
+  desyncDebug_ = dist_debug_level_ >= DistributedDebugLevel::INFO ? true : false;
```
could this just be `desyncDebug_ = dist_debug_level_ >= DistributedDebugLevel::INFO`?

kwen2501 replied:
Good idea

rohan-varma commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -540,7 +537,12 @@ ProcessGroupNCCL::ProcessGroupNCCL(
+  if (!desyncDebug_) {
+    // Provide backward support of NCCL_DESYNC_DEBUG
+    desyncDebug_ = parseEnvVarFlag(NCCL_DESYNC_DEBUG);
+  }
```
`desyncDebug_ = dist_debug_level_ >= DistributedDebugLevel::INFO || parseEnvVarFlag(NCCL_DESYNC_DEBUG)` to do it in one go?

kwen2501 replied:
Yes, will merge

rohan-varma reviewed (approved):
Nice work! Can we also update `distributed.rst` documentation to indicate that TDD >= INFO turns on nccl desync debug, and document the basic offerings of nccl_desync_debug? Or file an issue to do this.

kwen2501 committed:
[PyTorch Distributed] Consolidate NCCL_DESYNC_DEBUG and TORCH_DISTRIBUTED_DEBUG=INFO (#73257)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/73257

Infer desync debug from whether TORCH_DISTRIBUTED_DEBUG >= INFO

Test Plan:
1. When TORCH_DISTRIBUTED_DEBUG=INFO:
1.1 Catch mismatched collectives (e.g. broadcast vs reduce) - passed
1.2 Catch mismatched collective sizes - passed
2. QPS test: no performance regression - passed

Reviewed By: rohan-varma

Differential Revision: D34232827

fbshipit-source-id: 4cba47e22405ad5cfedf8241d961160485df8d6f
